numpy/random/_generator.pyx

```diff
@@ -4114,7 +4172,122 @@ cdef class Generator:
 
         return diric
 
-    # Shuffling and permutations:
+    def permuted(self, object x, *, axis=None, out=None):
+        """
+        permuted(x, axis=None, out=None)
+
+        Randomly permute `x` along axis `axis`.
+
+        Unlike `shuffle`, each slice along the given axis is shuffled
+        independently of the others.
+
+        Parameters
+        ----------
+        x : array_like
+            Array to be shuffled.
+        axis : int, optional
+            Slices of `x` in this axis are shuffled. Each slice
+            is shuffled independently of the others.  If `axis` is
+            None, the flattened array is shuffled.
+        out : ndarray, optional
+            If given, this is the destinaton of the shuffled array.
+            If `out` is None, a shuffled copy of the array is returned.
+
+        Returns
+        -------
+        ndarray
+            If `out` is None, a shuffled copy of `x` is returned.
+            Otherwise, the shuffled array is stored in `out`,
+            and `out` is returned
+
+        See Also
+        --------
+        shuffle
+        permutation
+
+        Examples
+        --------
+        Create a `numpy.random.Generator` instance:
+
+        >>> rng = np.random.default_rng()
+
+        Create a test array:
+
+        >>> x = np.arange(24).reshape(3, 8)
+        >>> x
+        array([[ 0,  1,  2,  3,  4,  5,  6,  7],
+               [ 8,  9, 10, 11, 12, 13, 14, 15],
+               [16, 17, 18, 19, 20, 21, 22, 23]])
+
+        Shuffle the rows of `x`:
+
+        >>> y = rng.permuted(x, axis=1)
+        >>> y
+        array([[ 4,  3,  6,  7,  1,  2,  5,  0],  # random
+               [15, 10, 14,  9, 12, 11,  8, 13],
+               [17, 16, 20, 21, 18, 22, 23, 19]])
+
+        `x` has not been modified:
+
+        >>> x
+        array([[ 0,  1,  2,  3,  4,  5,  6,  7],
+               [ 8,  9, 10, 11, 12, 13, 14, 15],
+               [16, 17, 18, 19, 20, 21, 22, 23]])
+
+        To shuffle the rows of `x` in-place, pass `x` as the `out`
+        parameter:
+
+        >>> y = rng.permuted(x, axis=1, out=x)
+        >>> x
+        array([[ 3,  0,  4,  7,  1,  6,  2,  5],  # random
+               [ 8, 14, 13,  9, 12, 11, 15, 10],
+               [17, 18, 16, 22, 19, 23, 20, 21]])
+
+        Note that when the ``out`` parameter is given, the return
+        value is ``out``:
+
+        >>> y is x
+        True
+        """
+
+        cdef int ax
+        cdef np.npy_intp axlen, axstride, itemsize
+        cdef void *buf
+        cdef np.flatiter it
+
+        x = np.asarray(x)
+
+        if out is None:
+            out = x.copy()
```

Comment:
This feel wrong.  If x is a view I think you want to return a non-strided C contiguous array in all cases.  I'd rather out = np.empty(x.shape, dtype=x.dtype)

Reply:
Not sure I follow, but I would agree that `x.copy("K")` would be better, since I think that is not the default.

Reply:
`out = np.empty(x.shape, dtype=x.dtype)` is probably closer to what the ufunc machinery does, so I'm inclined to agree. It also reduces the chance of subclass weirdness

Reply:
Seems like a copy of a non-contiguous array is C contiguous. A copy of a Fortran contiguous array is also C, not Fortran, contiguous.  So perhaps this is OK.

Reply:
@eric-wieser the ufunc machinery preserves memory order in such a case (it does that for optimization, but that is the result):
```
In [2]: (-np.ones((3,3), order="F")).flags.f_contiguous                                                                 
Out[2]: True
```

Reply:
In the case where `out = x.copy()` is run, we are creating a *new* array to hold the result.  I would expect this to be a C-contiguous array (which I believe is the default behavior of any other array creation routine).  That is the default behavior of the `copy()` method, so is there really a problem here?

Reply:
I honestly do not care too much either way (and I am not sure if `np.random` has a common pattern). In general, I would expect that operations on arrays tend to preserve the input contiguity. I.e. if all arrays involved are F-contiguous, the output should strive to be (ufuncs do this).

But: NumPy also often is purely C-order, and in general I would not have too much squirms about changing this even later. So if I had to vote, I slightly tend to `"K"` right now, but if you have doubts, C – the default for the copy method, is the safe bet.

Reply:
My expectation would be K too, but not strongly.

Reply:
Oh, but the point of @seberg's comment was that ufuncs preserve memory order, so, for exmple, `-x` and `sin(x)` have the same memory layout as `x`.  My expectation for this function matches @bashtage's: I expected this to create a new array with C-contiguous memory layout.  Is it really preferable to maintain the memory layout of the input? In this case,  the optimization argument doesn't apply, and the function does not implement an element-wise operation, so maintaining the layout of the input does not seem important.

Reply:
(In case my last comment seems worded strangely: I was writing my previous comment before @seberg's and @eric-wieser's most recent comments, but clicked the "Comment" button after them.)

Reply:
@WarrenWeckesser but for example `np.copy()` does default to `"K"` (it differs from the method ☹). We are not very consistent about it, but I think even if we consider this an array-creation function, I would tend towards K.

Reply:
OK, `K` it is.